src/components/MoneyRequestAmountInput.tsx

```diff
@@ -0,0 +1,251 @@
+            start: frontendAmount.length,
+            end: frontendAmount.length,
+        });
+        // we want to re-initialize the state only when the amount changes
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [amount]);
+
+    // Modifies the amount to match the decimals for changed currency.
+    useEffect(() => {
+        // If the changed currency supports decimals, we can return
+        if (MoneyRequestUtils.validateAmount(currentAmount, decimals)) {
+            return;
+        }
+
+        // If the changed currency doesn't support decimals, we can strip the decimals
+        setNewAmount(MoneyRequestUtils.stripDecimalsFromAmount(currentAmount));
+
+        // we want to update only when decimals change (setNewAmount also changes when decimals change).
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [setNewAmount]);
+
+    /**
+     * Input handler to check for a forward-delete key (or keyboard shortcut) press.
+     */
+    const textInputKeyPress = ({nativeEvent}: NativeSyntheticEvent<KeyboardEvent>) => {
+        const key = nativeEvent?.key.toLowerCase();
+        if (Browser.isMobileSafari() && key === CONST.PLATFORM_SPECIFIC_KEYS.CTRL.DEFAULT) {
+            // Optimistically anticipate forward-delete on iOS Safari (in cases where the Mac Accessiblity keyboard is being
+            // used for input). If the Control-D shortcut doesn't get sent, the ref will still be reset on the next key press.
+            forwardDeletePressedRef.current = true;
+            return;
+        }
+        // Control-D on Mac is a keyboard shortcut for forward-delete. See https://support.apple.com/en-us/HT201236 for Mac keyboard shortcuts.
+        // Also check for the keyboard shortcut on iOS in cases where a hardware keyboard may be connected to the device.
+        const operatingSystem = getOperatingSystem();
+        forwardDeletePressedRef.current = key === 'delete' || ((operatingSystem === CONST.OS.MAC_OS || operatingSystem === CONST.OS.IOS) && nativeEvent?.ctrlKey && key === 'd');
+    };
+
+    const formattedAmount = MoneyRequestUtils.replaceAllDigits(currentAmount, toLocaleDigit);
+
+    return (
+        <TextInputWithCurrencySymbol
+            formattedAmount={formattedAmount}
+            onChangeAmount={setNewAmount}
+            onCurrencyButtonPress={onCurrencyButtonPress}
+            placeholder={numberFormat(0)}
+            ref={(ref) => {
+                if (typeof forwardedRef === 'function') {
+                    forwardedRef(ref);
+                } else if (forwardedRef?.current) {
+                    // eslint-disable-next-line no-param-reassign
+                    forwardedRef.current = ref;
+                }
+                textInput.current = ref;
+            }}
+            selectedCurrencyCode={currency}
+            selection={selection}
+            onSelectionChange={(e: NativeSyntheticEvent<TextInputSelectionChangeEventData>) => {
```

Comment:
Coming from this issue https://github.com/Expensify/App/issues/41762 , we are encountering an edge case on Android. When we select all the number and then add a number, the cursor is misplaced and appears at the start of the number. The reason for this is that on Android, when we select all the text and add a number, the selection update from onSelectionChange is triggered earlier than the manual selection update. As a result, the prevSelection already contains the new selection, leading to an incorrect selection calculation more context on the proposal https://github.com/Expensify/App/issues/41762#issuecomment-2104107478